### ISSUE TITLE
resolve "~" to home directory in read_csv and possibly other read methods

### ISSUE THREAD

**Issue by marcvanheerden** [pola-rs/polars#882]
It seems that the "~" character is not resolving to the user's home directory in the path input for pl.read_csv.

I can have a look at getting this to work for read_csv and the other read methods.

```
Python 3.9.5 (default, May 18 2021, 12:31:01)
[Clang 10.0.0 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.

$ import polars as pl

$ import pandas as pd

$ path1 = "~/Desktop/test.csv"

$ path2 = "/Users/41745/Desktop/test.csv"

$ df = pl.read_csv(path1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/anaconda3/envs/polarstest/lib/python3.9/site-packages/polars/functions.py", line 202, in read_csv
    tbl = pa.csv.read_csv(
  File "pyarrow/_csv.pyx", line 793, in pyarrow._csv.read_csv
  File "pyarrow/_csv.pyx", line 686, in pyarrow._csv._get_reader
  File "pyarrow/io.pxi", line 1493, in pyarrow.lib.get_input_stream
  File "pyarrow/io.pxi", line 1464, in pyarrow.lib.get_native_file
  File "pyarrow/io.pxi", line 827, in pyarrow.lib.OSFile.__cinit__
  File "pyarrow/io.pxi", line 837, in pyarrow.lib.OSFile._open_readable
  File "pyarrow/error.pxi", line 141, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/error.pxi", line 110, in pyarrow.lib.check_status
FileNotFoundError: [Errno 2] Failed to open local file '~/Desktop/test.csv'. Detail: [errno 2] No such file or directory

$ df = pl.read_csv(path2)

$ df2 = pd.read_csv(path1)
sys:1: DtypeWarning: Columns (4) have mixed types.Specify dtype option on import or set low_memory=False.

$
```

**Comment by olivier-lacroix** [comment]
Hi, I have just opened a PR (https://github.com/pola-rs/polars/pull/883) which defers file opening to fsspec. I _believe_ fsspec handles "~" in paths

**Comment by ritchie46** [comment]
@marcvanheerden , I think this might be an easy addition in the parsers.

**Comment by marcvanheerden** [comment]
Will do

**Comment by marcvanheerden** [comment]
going to wait and see what happens with #883 before continuing here

**Comment by ritchie46** [comment]
Right. I think it does remain a good addition though. As #883 will only fix things from the python side.

**Comment by marcvanheerden** [comment]
Sorry for the back and forth here, do you think an upstream fix in arrow would be better?

**Comment by ritchie46** [comment]
I think in Polars itself. Polars uses a custom csv-parser and parsers from arrow. I think that in any code where we accept a path/string to a filepath, we could evaluate the absolute path before we call Polars/arrows parsers.

The fix in #833 is welcome, but I view those as a temporary fix, as this only influence the python bindings.
